src/document-watcher.js

```diff
@@ -71,19 +71,32 @@ function handler(mxns) {
         if (!host) {
           continue;
         }
+
         newScope = getIsExtends(host).is;
-        if (currentScope === newScope) {
-          // make sure all the subtree elements are scoped correctly
-          let unscoped = window['ShadyDOM']['nativeMethods']['querySelectorAll'].call(
-            n, `:not(.${StyleTransformer.SCOPE_NAME})`);
-          for (let j = 0; j < unscoped.length; j++) {
-            StyleTransformer.element(unscoped[j], currentScope);
+        // make sure all the subtree elements are scoped correctly
+        let unscoped = window['ShadyDOM']['nativeMethods']['querySelectorAll'].call(
+          n, `:not(.${StyleTransformer.SCOPE_NAME})`);
```

Comment:
As far as I can tell that's the core issue that I was running into. Because this is run in a mutationObserver when a large tree is inserted only the root of that tree will be passed into this function as an `addedNode` in the mutation list. Previously this was selecting everything and applying the root's scope, which is definitely incorrect since this goes past scope boundaries. So instead of updating all unscoped nodes to the root's scope this grabs all unscoped nodes, finds their correct scope, and scopes them.

I am concerned though about the fact that we're digging so deeply into the tree as you're saying. It seems dangerous to grab all of the nodes in the tree like this and run over them. I *think* the safety comes in that we're only grabbing nodes that don't have `.style-scope` on them so we won't touch anything that's already been scoped. And those that haven't been scoped should be scoped.

Reply:
Oh I see now. Yes `getIsExtends(unscopedHost)` makes more sense now. Maybe we can improve the naming a little bit to clarify?

What about:
`s/unscoped/usncopedNodes/`
`s/unscopedRoot/rootForUnscopedNode/`
`s/unscopedHost/hostForUnscopedNode/`
`s/newScope/scopeForPreviouslyUnscopedNode/`

And then maybe add another comment saying:
```
To make sure that any unscoped nodes that were inserted in the current batch are correctly styled,
query all unscoped nodes and force their style-scope to be applied.
This could happen if a sub-element appended an unscoped node in its shadowroot and this function
runs on a parent element of the host of that unscoped node:
parent-element -> element -> unscoped node
Here unscoped node should have the style-scope element, not parent-element.
```